### PR TITLE
[wasm] deprecate legacy JS API and propose new

### PR DESCRIPTION
**Pull request by pavelsavara**
- Added new API methods to top level next to `MONO` and `BINDING` namespaces
    - `runMain`, `runMainAndExit`, `setEnvironmentVariable`, `getAssemblyExports` are just renamed existing API
    - `getConfig` will return `mono-config.json` type `MonoConfig`
    -  `setHeapI64Big`, `setHeapI32` , `getHeapI32`, ... memory functions
- marked `MONO` and `BINDING` namespaces obsolete
- separated legacy API into `dotnet-legacy.d.ts`
- renamed `snake_case` to `camelCase` names 
    - the idea is that these types were mostly added to Net7 and we have still chance to fix it without backward compatibility shim. 
    - The shim could be done, if really necessary.

## Details
- in the `mono-config.json`/`MonoConfig`type 
    - `assembly_root` -> `assemblyRootFolder`
    - `debug_level` -> `debugLevel`
    - `enable_debugging` - removed in favor of `debugLevel`
    - `isError` - removed, we throw exceptions as necessary
    - `globalization_mode` -> `globalizationMode`
    - `diagnostic_tracing` -> `diagnosticTracing`
    - `remote_sources` -> `remoteSources`
    - `environment_variables` -> `environmentVariables`
    - `runtime_options` -> `runtimeOptions`
    - `aot_profiler_options` -> `aotProfilerOptions` with `writeAt` & `sendTo`
    - `coverage_profiler_options` -> `coverageProfilerOptions` with `writeAt` & `sendTo`
    - `diagnostic_options` -> `diagnosticOptions`
    - `ignore_pdb_load_errors` -> `ignorePdbLoadErrors`
    - `wait_for_debugger` -> `waitForDebugger`
- in `AssetEntry`
    - `virtual_path` -> `virtualPath`
    - `load_remote` -> `loadRemote`
    - `is_optional` -> `isOptional`
    - we are producing it in `src\tasks\WasmAppBuilder\WasmAppBuilder.cs`
- in `DiagnosticServerOptions`
    - `connect_url` -> `connectUrl`
- in `UnnamedProviderConfiguration`
    - `keyword_mask` -> `keywordMask`
- in `Diagnostics`
    - `EventLevel` -> `eventLevel`
- updated samples and templates
- renamed runtime arg `--diagnostic_tracing` -> `--diagnostic-tracing`
- replaced `IMPORTS` with `setModuleImports`
- renamed `runtimeBuildInfo` & `productVersion` & `buildConfiguration` to first lowercase

### CODE REVIEW

File: src/mono/sample/wasm/browser/main.js [hunk 1]
```diff
@@ -1,16 +1,11 @@
+// @ts-check
+// @ts-ignore
 import createDotnetRuntime from './dotnet.js'
 
-function wasm_exit(exit_code, reason) {
```
[the remark this answers is not in the document]
Comment (pavelsavara) [reply src/mono/sample/wasm/browser/main.js:3]: okay, reverted.

File: src/mono/wasm/runtime/export-api.ts [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+import { APIType } from "./exports";
+import { runtimeHelpers } from "./imports";
+import { mono_wasm_get_assembly_exports } from "./invoke-cs";
+import { getB32, getF32, getF64, getI16, getI32, getI52, getI64Big, getI8, getU16, getU32, getU52, getU8, setB32, setF32, setF64, setI16, setI32, setI52, setI64Big, setI8, setU16, setU32, setU52, setU8 } from "./memory";
+import { mono_run_main, mono_run_main_and_exit } from "./run";
+import { mono_wasm_setenv } from "./startup";
+import { MonoConfig } from "./types";
+
+export function export_api(): any {
+    const api: APIType = {
```
[the remark this answers is not in the document]
Comment (pavelsavara) [reply src/mono/wasm/runtime/export-api.ts:11]: there are only few methods right now and it's hard to find good category names.

**Review by maraf — APPROVED** [review]

**Review by maraf — CHANGES_REQUESTED** [review]
I think we could remove `EXPORTS` and `IMPORTS`. I like the API for getting assembly exports (`getAssemblyExports`) and as imports are similar feature, we could provide a similar API for it, eg. `setAssemblyImports(assembly: string, imports: any)`.
With this API we can remove `EXPORTS` and `IMPORTS`.

I'm still not sure if flatten `API` object is the right choice. I know than by now we have only few functions, but it seems to me that adding more structure is good for future enhancements.

**Comment by maraf** [comment]
After a discussion we have chosen a 100% flat API in root object returned from the `createDotnetRuntime`.

File: src/mono/wasm/runtime/es6/dotnet.es6.lib.js [hunk 3]
```diff
@@ -33,7 +33,7 @@ if (ENVIRONMENT_IS_NODE) {
 }
 let __dotnet_exportedAPI = __dotnet_runtime.__initializeImportsAndExports(
     { isGlobal:false, isNode:ENVIRONMENT_IS_NODE, isWorker:ENVIRONMENT_IS_WORKER, isShell:ENVIRONMENT_IS_SHELL, isWeb:ENVIRONMENT_IS_WEB, isPThread:${isPThread}, quit_, ExitStatus, requirePromise:__dotnet_replacements.requirePromise },
-    { mono:MONO, binding:BINDING, internal:INTERNAL, module:Module, marshaled_exports: EXPORTS, marshaled_imports: IMPORTS },
+    { api:API, mono:MONO, binding:BINDING, internal:INTERNAL, module:Module, marshaled_imports: IMPORTS },
```
Comment (maraf) [inline src/mono/wasm/runtime/es6/dotnet.es6.lib.js:36]: This needs a change, as the API object doesn't exist anymore, right?

File: src/mono/wasm/runtime/es6/dotnet.es6.pre.js [hunk 4]
```diff
@@ -1,10 +1,10 @@
-const MONO = {}, BINDING = {}, INTERNAL = {}, IMPORTS = {}, EXPORTS = {};
+const API = {}, MONO = {}, BINDING = {}, INTERNAL = {}, IMPORTS = {};
```
Comment (maraf) [inline src/mono/wasm/runtime/es6/dotnet.es6.pre.js:1]: This needs a change, as the API object doesn't exist anymore, right?

**Review by maraf — APPROVED** [review]

**Review by ilonatommy — APPROVED** [review]

File: src/mono/sample/wasm/browser/main.js [hunk 5]
```diff
@@ -12,43 +12,67 @@ function wasm_exit(exit_code, reason) {
...
+                // config is loaded and could be tweaked before the rest of the runtime startup sequence
+                config.environmentVariables["MONO_LOG_LEVEL"] = "debug"
+            },
             preInit: () => { console.log('user code Module.preInit'); },
             preRun: () => { console.log('user code Module.preRun'); },
-            onRuntimeInitialized: () => { console.log('user code Module.onRuntimeInitialized'); },
+            onRuntimeInitialized: () => {
+                console.log('user code Module.onRuntimeInitialized');
+                // here we could use API passed into this callback
+                // Module.FS.chdir("/");
```
[the remark this answers is not in the document]
Comment (pavelsavara) [reply src/mono/sample/wasm/browser/main.js:43]: Yes, this is just sample code we would not like to run

File: src/mono/wasm/runtime/types.ts [hunk 6]
```diff
@@ -331,8 +324,7 @@ export type EarlyExports = {
     binding: any,
     internal: any,
     module: any,
-    marshaled_exports: any,
-    marshaled_imports: any
+    marshaled_imports: any,
```
[the remark this answers is not in the document]
Comment (pavelsavara) [reply src/mono/wasm/runtime/types.ts:327]: I only changed the public identifiers. We could re-format the rest of typeScript codebase in Net8

**Review by lambdageek — APPROVED** [review]

**Comment by pavelsavara** [comment]
`System.NotSupportedException : The deserialization constructor for type 'System.Tuple`1[System.Int32]' contains parameters with null names.` is https://github.com/dotnet/runtime/issues/73347

Debugger failures are https://github.com/dotnet/runtime/issues/73342

**Pull request merged**